chorus_lib/src/transport/local.rs

```diff
@@ -5,52 +5,107 @@ use std::sync::Arc;
 
 use serde_json;
 
-use crate::core::{Portable, Transport};
+use std::marker::PhantomData;
+
+use crate::LocationSet;
+
+use crate::core::{ChoreographyLocation, HCons, HList, Portable, Transport};
 use crate::utils::queue::BlockingQueue;
 
 type QueueMap = HashMap<String, HashMap<String, BlockingQueue<String>>>;
 
-/// The local transport.
-///
-/// This transport uses a blocking queue to allow for communication between threads. Each location must be executed in its thread.
-///
-/// Unlike network-based transports, all locations must share the same `LocalTransport` instance. The struct implements `Clone` so that it can be shared across threads.
-#[derive(Clone)]
-pub struct LocalTransport {
-    internal_locations: Vec<String>,
+/// A Transport channel used between multiple `Transport`s.
+pub struct LocalTransportChannel<L: HList> {
+    /// The location set where the channel is defined on.
+    location_set: std::marker::PhantomData<L>,
     queue_map: Arc<QueueMap>,
 }
 
-impl LocalTransport {
-    /// Creates a new `LocalTransport` instance from a list of locations.
-    pub fn from(locations: &[&str]) -> Self {
+impl<L: HList> Clone for LocalTransportChannel<L> {
+    fn clone(&self) -> Self {
+        LocalTransportChannel {
+            location_set: PhantomData,
+            queue_map: self.queue_map.clone(),
+        }
+    }
+}
+
+impl LocalTransportChannel<LocationSet!()> {
+    /// Creates a new `LocalTransportChannel` instance
+    pub fn new() -> Self {
+        Self {
+            location_set: PhantomData,
+            queue_map: HashMap::new().into(),
+        }
+    }
+}
+
+impl<L: HList> LocalTransportChannel<L> {
+    /// Adds a new location to the set of locations in the `LocalTransportChannel`.
+    pub fn with<NewLocation: ChoreographyLocation>(
+        self,
+        _location: NewLocation,
+    ) -> LocalTransportChannel<HCons<NewLocation, L>> {
         let mut queue_map: QueueMap = HashMap::new();
-        for sender in locations.clone() {
+        let mut str_list = L::to_string_list();
+        str_list.push(NewLocation::name());
+        for sender in &str_list {
             let mut n = HashMap::new();
-            for receiver in locations.clone() {
+            for receiver in &str_list {
                 n.insert(receiver.to_string(), BlockingQueue::new());
             }
             queue_map.insert(sender.to_string(), n);
         }
+
+        LocalTransportChannel {
+            location_set: PhantomData,
+            queue_map: queue_map.into(),
+        }
+    }
+}
+
+/// The local transport.
+///
+/// This transport uses a blocking queue to allow for communication between threads. Each location must be executed in its thread.
+///
+/// Unlike network-based transports, all locations must share the same `LocalTransport` instance. The struct implements `Clone` so that it can be shared across threads.
```

Comment:
I think this comment is no longer true?

```suggestion
/// Unlike network-based transports, all locations must share the same `LocalTransportChannel` instance. 
```

Reply:
I do agree with your point, BUT: `LocalTransportChannel` is not a field in network-based transports, so it doesn't even make sense to compare the two. Can we just go with "all locations must share the same `LocalTransportChannel` instance"?